File: api/src/main/java/fr/huiitre/tools/modules/riot/sync/infrastructure/PostgresValorantWeaponSyncRepository.java
package fr.huiitre.tools.modules.riot.sync.infrastructure;

import java.util.List;
import java.util.UUID;

import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.core.RowMapper;

import fr.huiitre.tools.modules.riot.sync.application.ValorantWeaponSyncData;
import fr.huiitre.tools.modules.riot.sync.application.ValorantWeaponSyncRepository;
import fr.huiitre.tools.modules.riot.valorant.application.view.ValorantWeaponView;

public class PostgresValorantWeaponSyncRepository implements ValorantWeaponSyncRepository {

    private final JdbcTemplate jdbcTemplate;

    public PostgresValorantWeaponSyncRepository(JdbcTemplate jdbcTemplate) {
        this.jdbcTemplate = jdbcTemplate;
    }

    private static final RowMapper<ValorantWeaponView> ROW_MAPPER = (rs, rowNum) -> new ValorantWeaponView(
            rs.getLong("id"),
            rs.getObject("asset_id", UUID.class),
            rs.getString("name"),
            rs.getString("category"),
            rs.getObject("default_skin_asset_id", UUID.class),
            rs.getString("display_icon_url"));

    @Override
    public List<ValorantWeaponView> findAll() {
        final String sql = """
                SELECT id, asset_id, name, category, default_skin_asset_id, display_icon_url
                FROM tools_riot.valorant_weapons
                """;
        return jdbcTemplate.query(sql, ROW_MAPPER);
    }

    @Override
    public Long save(ValorantWeaponSyncData data) {
        final String sql = """
                INSERT INTO tools_riot.valorant_weapons (asset_id, name, category, default_skin_asset_id, display_icon_url)
                VALUES (?, ?, ?, ?, ?)
                RETURNING id
                """;
        return jdbcTemplate.queryForObject(sql, Long.class,
                data.getAssetId(),
                data.getName(),
                data.getCategory(),
                data.getDefaultSkinAssetId(),
                data.getDisplayIconUrl());
    }

    @Override
    public void update(Long id, ValorantWeaponSyncData data) {
        final String sql = """
                UPDATE tools_riot.valorant_weapons
                SET name = ?, category = ?, default_skin_asset_id = ?, display_icon_url = ?, updated_at = now()
                WHERE id = ?
                """;
        jdbcTemplate.update(sql,
                data.getName(),
                data.getCategory(),
                data.getDefaultSkinAssetId(),
                data.getDisplayIconUrl(),
                id);
    }

    @Override
    public void delete(Long id) {
        jdbcTemplate.update("DELETE FROM tools_riot.valorant_weapons WHERE id = ?", id);
    }
}

File: api/Modules/Riot/Sync/Infrastructure/PostgresValorantWeaponSyncRepository.cs
using Tools.Api.Modules.Riot.Sync.Application;
using Tools.Api.Modules.Riot.Common.Infrastructure;
using Tools.Api.Modules.Riot.Sync.Application.Ports;
using Tools.Api.Modules.Riot.Valorant.Application.Catalog.Views;

namespace Tools.Api.Modules.Riot.Sync.Infrastructure;

public sealed class PostgresValorantWeaponSyncRepository(RiotDatabase database) : IValorantWeaponSyncRepository
{
    public Task<List<ValorantWeaponView>> FindAll() =>
        database.Query<ValorantWeaponView>(
            """
            SELECT id AS Id, asset_id AS AssetId, name AS Name, category AS Category,
                   default_skin_asset_id AS DefaultSkinAssetId, display_icon_url AS DisplayIconUrl
            FROM tools_riot.valorant_weapons
            """);

    public Task<long> Save(ValorantWeaponSyncData data) =>
        database.ExecuteScalar<long>(
            """
            INSERT INTO tools_riot.valorant_weapons (asset_id, name, category, default_skin_asset_id, display_icon_url)
            VALUES (@AssetId, @Name, @Category, @DefaultSkinAssetId, @DisplayIconUrl)
            RETURNING id
            """,
            data);

    public Task Update(long id, ValorantWeaponSyncData data) =>
        database.Execute(
            """
            UPDATE tools_riot.valorant_weapons
            SET name = @Name, category = @Category, default_skin_asset_id = @DefaultSkinAssetId,
                display_icon_url = @DisplayIconUrl
            WHERE id = @Id
            """,
            new { Id = id, data.Name, data.Category, data.DefaultSkinAssetId, data.DisplayIconUrl });

    public Task Delete(long id) =>
        database.Execute("DELETE FROM tools_riot.valorant_weapons WHERE id = @Id", new { Id = id });
}
